test(build-tools): add glob and gitignore behavior tests

tylerbutler opened the pull request:
## Summary

Adds comprehensive tests for glob and gitignore behavior in build-tools to de-risk future migration from  to .

## Changes

- Extract  function from  into  for testability
- Add tests for  wrapper covering wildcards, globstar, , , , , and  options
- Add tests for  covering gitignore integration, pattern matching, and directory exclusion
- Add runtime order randomization for glob tests to ensure consistent cross-library behavior
- Add test data files for gitignore pattern testing

## Why

These tests document current glob behavior and will verify that the upcoming  migration preserves existing functionality.

tylerbutler committed:
test(build-tools): add tests for glob-related functions

tylerbutler committed:
Add tests for glob and gitignore behavior before dependency migration

This change adds tests to reduce risk when implementing the glob library
consolidation described in plans/dependency-reduction.md.

Changes:
- Extract globWithGitignore function from LeafTask for testability
- Add comprehensive tests for globFn wrapper (nodir, dot, ignore, cwd, absolute options)
- Add tests for globWithGitignore gitignore behavior
- Add test data files for gitignore pattern testing

The globWithGitignore function encapsulates the globby usage with gitignore
support, making it easier to migrate to tinyglobby in the future. When
migrating, only this function needs to change, and the tests will verify
the behavior is preserved.

tylerbutler committed:
add header

tylerbutler committed:
rename

tylerbutler committed:
Merge branch 'main' into glob-tests

tylerbutler committed:
Merge branch 'main' into glob-tests

tylerbutler committed:
Fix failing tests

- Fix globPatterns.test.ts: Update glob pattern from *.js to *.mjs to match actual test file
- Fix globPatterns.test.ts: Add beforeEach/afterEach hooks to dynamically create/cleanup gitignored test files
- Fix canonicalizeChangesets.test.ts: Serialize git operations to avoid index.lock race conditions

All 703 tests now passing.

tylerbutler committed:
Merge branch 'main' into glob-tests

tylerbutler commented on build-tools/packages/build-cli/src/test/library/canonicalizeChangesets.test.ts:
```diff
@@ -210,17 +210,14 @@ describe("canonicalizeChangesets", () => {
+				`---\n"@fluid/package-${i}": patch\n"__custom": "metadata"\n---\n\nChange ${i}\n\nDetailed description.\n`,
 			);
 		}
-		await Promise.all(writePromises);
```
Are these changes related to the others? This seems to completely change how the test functions.

tylerbutler committed:
Revert unrelated test change that altered test semantics

alexvy86 commented on build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/leafTask.ts:
```diff
@@ -738,13 +738,10 @@ export abstract class LeafWithGlobInputOutputDoneFileTask extends LeafWithFileSt
-		const files = await globby(globs, {
+		return globWithGitignore(globs, {
 			cwd: this.node.pkg.directory,
-			// file paths returned from getInputFiles and getOutputFiles should always be absolute
```
Something similar to this comment feels worth moving to the docs of `globWithGitIgnore` to makes its expectations clear, probably the docs of its `patterns` parameter. And/or make the behavior controllable with `GlobWithGitIgnoreOptions`.

alexvy86 commented on build-tools/packages/build-tools/src/test/data/glob/.gitignore:
```diff
@@ -0,0 +1,3 @@
+# Test gitignore
```
```suggestion
# Test gitignore
# The entries here don't exist in the repo, they are created by tests when they run.
```

alexvy86 commented on build-tools/packages/build-tools/src/test/globPatterns.test.ts:
```diff
@@ -0,0 +1,276 @@
+		assert(
+			!filenames.includes("test.ignored"),
+			"Should exclude files matching *.ignored pattern",
+		);
```
Nothing creates the `test.ignored` file as far as I can tell so this test passes without actually validating anything.

tylerbutler committed:
Address PR feedback for glob tests

- Document globWithGitignore pattern expectations in JSDoc
- Remove migration-specific language from test docstrings
- Use exact list assertions (deepEqual) for all tests
- Simplify toPosixPath to always convert backslashes
- Update .gitignore comment per reviewer suggestion
- Create test.ignored dynamically in beforeEach with existence assertions

tylerbutler committed:
Update build-tools/packages/build-tools/src/test/data/glob/.gitignore

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

tylerbutler commented on build-tools/packages/build-tools/src/test/globPatterns.test.ts:
```diff
@@ -0,0 +1,341 @@
+
+		it("returns relative paths when absolute is not set", async () => {
+			const results = await globFn("*.ts", { cwd: globTestDataPath, nodir: true });
+			for (const result of results) {
```
Again assert the contents.

tylerbutler commented on build-tools/packages/build-tools/src/test/globPatterns.test.ts:
```diff
@@ -0,0 +1,341 @@
+		const result = toPosixPath(input);
+		assert.equal(result, "src/utils/file.ts");
+	});
+});
```
Add tests for absolute paths in both POSIX and Windows format.

tylerbutler reviewed (commented):
Biggest issues are:

- The `const filenames = results.map((f) => path.basename(f)).sort();` pattern is not ideal for a lot of tests. We should be able to include more of the path than just the basename, and the sorting seems unneeded - or is the order not guaranteed? Either way we should clarify that.
- Other tests verify absolute vs relative paths, but they don't check contents but should.
- Maybe some test helpers would make the tests more DRY.

tylerbutler committed:
Merge branch 'main' into glob-tests

tylerbutler committed:
test(build-tools): improve glob test assertions and reduce duplication

Address PR feedback:
- Add toRelativePaths() helper to show full path structure instead of just basenames
- Add assertAbsolutePaths() helper to verify both path type and contents
- Update all tests to use relative paths for better test clarity
- Tests now verify the complete path structure, not just filenames
- Remove duplicate comment line in .gitignore

This makes tests more maintainable and easier to understand what files
are actually being matched by the glob patterns.

tylerbutler reviewed (commented):
## Addressed PR feedback

✅ **Fixed test assertions** - Replaced `path.basename()` pattern with:
- `toRelativePaths()` helper that shows full path structure (e.g., "nested/dir1/file1.ts" instead of just "file1.ts")
- This makes tests much clearer about which files are being matched

✅ **Improved absolute path tests** - Added `assertAbsolutePaths()` helper that:
- Verifies paths are absolute
- Also checks the basenames to confirm correct files were matched

✅ **Made tests more DRY** - Added two helper functions to reduce repetition across all test cases

✅ **Removed duplicate .gitignore comment**

All tests pass with these improvements. The relative paths now make it clear which files in which directories are being matched, eliminating the ambiguity from when we only checked basenames.

tylerbutler committed:
test(build-tools): add ordering behavior tests for glob libraries

Add tests to document and verify the ordering behavior of glob libraries:

- globFn (using 'glob' package): Returns results in sorted order by default
- globWithGitignore (using 'globby'): Returns results in breadth-first order
  (root files before subdirectory files), NOT lexicographically sorted

These tests will alert us if ordering behavior changes when migrating
to alternative glob libraries like tinyglobby. This is why we explicitly
sort results in other tests - to ensure deterministic assertions regardless
of library implementation details.

Also fixed test expectations to match actual test data directory structure
(nested/deep/ instead of nested/dir1/, nested/dir2/).

tylerbutler committed:
feat(build-tools): add runtime order randomization for glob testing

Implement FLUID_BUILD_TEST_RANDOM_ORDER environment variable to shuffle
glob results and expose code that incorrectly depends on glob ordering.

Changes:
- Add shuffleArray() utility using Fisher-Yates algorithm
- Add isRandomOrderTestMode() to check environment variable
- Modify globFn() to shuffle results when test mode is enabled
- Modify globWithGitignore() to shuffle results when test mode is enabled
- Add comprehensive tests for randomization behavior
- Add documentation in docs/TESTING_ORDER_INDEPENDENCE.md

Usage:
  FLUID_BUILD_TEST_RANDOM_ORDER=true npm run build

This will catch bugs like:
- JSON.stringify() on unsorted file arrays
- Building objects from glob results (insertion order matters)
- Any code assuming specific glob result ordering

When migrating to alternative glob libraries (e.g., tinyglobby), running
builds with this flag will immediately expose any ordering dependencies
that would break with different library implementations.

See docs/TESTING_ORDER_INDEPENDENCE.md for full documentation and examples.

tylerbutler committed:
rm doc

tylerbutler reviewed (commented):
Resolving feedback items that have been addressed in the latest commits:

✅ **Comment 2585817184** - Added clarifying comment to .gitignore file
✅ **Comment 2585803578** - Fixed toPosixPath test to use hardcoded expected output
✅ **Comment 2585737653** - All tests now use assert.deepEqual with exact expected arrays
✅ **Comment 2586057086** - Made tests more DRY with helper functions (toRelativePaths, assertAbsolutePaths)
✅ **Comment 2586050687** - Absolute path test now asserts contents
✅ **Comment 2586051572** - Relative path test now asserts exact contents
✅ **Comment 2585823923** - test.ignored file is now created in beforeEach
✅ **Comment 2586044900 & 2586046225** - Added explicit ordering test to document glob library behavior

tylerbutler committed:
Address remaining PR feedback: add @privateRemarks to test docs and absolute path tests

tylerbutler committed:
Merge branch 'main' into glob-tests

alexvy86 commented on build-tools/packages/build-tools/src/fluidBuild/tasks/taskUtils.ts:
```diff
@@ -97,3 +118,75 @@ export async function loadModule(modulePath: string, moduleType?: string) {
+		[array[i], array[j]] = [array[j], array[i]];
+	}
+	return array;
+}
```
Super nit: since it shuffles in place, I think an API that doesn't return anything might prevent bugs where people assume the original array remains untouched because they got a new one.

tylerbutler commented on build-tools/packages/build-tools/src/test/globPatterns.test.ts:
```diff
@@ -0,0 +1,526 @@
+		assert(
+			!relativePaths.includes("gitignored/shouldBeIgnored.ts"),
+			"gitignored/shouldBeIgnored.ts should be excluded",
+		);
```
This is really meant to assert test setup went correctly sinc the files don't exist by default. I guess it could be in the before block, but I try to keep assertions out of there.

alexvy86 replied:
And that's the last thing, the rest of the PR looks good to me :).

tylerbutler replied:
Fixed in latest.

tylerbutler committed:
Apply suggestions from code review

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

tylerbutler committed:
feedback

alexvy86 reviewed (approved):
🚀

tylerbutler committed:
Merge branch 'main' into glob-tests
